### PR TITLE
soc: renesas: ra: ra4m1: Migrate to FSP-based configuration

### PR DESCRIPTION
**Pull request by soburi**
Change to use FSP to integrate with other Renesas RA series.

### CODE REVIEW

**Review by KhiemNguyenT — CHANGES_REQUESTED** [review]
@soburi , I think you need to brush up the PR. There are some changes for RA2A1 are mixed into.

**Comment by soburi** [comment]
> @soburi , I think you need to brush up the PR. There are some changes for RA2A1 are mixed into.

Update done.
I will remove DNM and proceed once the following PR in hal_renesas is approved.

https://github.com/zephyrproject-rtos/hal_renesas/pull/31

File: dts/arm/renesas/ra/ra-cm4-common.dtsi [hunk 1]
```diff
@@ -21,49 +22,51 @@
...
-			status = "disabled";
-			clocks = <&mosc>;
-			clock-div = <2>;
-			clock-mult = <8>;
+			compatible = "renesas,ra-cgc-pll";
 			#clock-cells = <0>;
+
+			/* PLL */
+			source = <RA_PLL_SOURCE_MAIN_OSC>;
+			div = <RA_PLL_DIV_2>;
```
Comment (gmarull) [inline dts/arm/renesas/ra/ra-cm4-common.dtsi:67]: wrong:

- source: define another clock and make it source using eg `clocks =  <&main_osc>;`
- divider should be given as integer `div = <2>;`.

File: dts/arm/renesas/ra/ra-cm4-common.dtsi [hunk 2]
```diff
@@ -82,19 +85,64 @@
...
+			iclk: iclk {
+				compatible = "renesas,ra-cgc-pclk";
+				clk_div = <RA_SYS_CLOCK_DIV_16>;
+				#clock-cells = <2>;
+				status = "okay";
+			};
+
+			pclka: pclka {
+				compatible = "renesas,ra-cgc-pclk";
+				clk_div = <RA_SYS_CLOCK_DIV_16>;
```
Comment (gmarull) [inline dts/arm/renesas/ra/ra-cm4-common.dtsi:108]: no underscores in properties please.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: dts/arm/renesas/ra/ra-cm4-common.dtsi (hunk 2 above)
Comment (soburi) [reply dts/arm/renesas/ra/ra-cm4-common.dtsi:108]: I have created https://github.com/zephyrproject-rtos/zephyr/pull/78097.
I will fix it after this PR is merged.

File: dts/arm/renesas/ra/ra-cm4-common.dtsi (hunk 1 above)
Comment (soburi) [reply dts/arm/renesas/ra/ra-cm4-common.dtsi:67]: @duynguyenxa
Are these properties actually used?

Comment (duynguyenxa) [reply dts/arm/renesas/ra/ra-cm4-common.dtsi:67]: @soburi , these properties are used to provide information for the bsp_clock_cfg.h in the hal layer:
```
#define BSP_CFG_PLL_SOURCE                                                                         \
	BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pll), source, RA_PLL_SOURCE_DISABLE)
#define BSP_CFG_PLL_DIV BSP_CLOCK_PROP_HAS_STATUS_OKAY_OR(DT_NODELABEL(pll), div, RA_PLL_DIV_2)
#if DT_NODE_HAS_STATUS(DT_NODELABEL(pll), okay)
```

The BSP_CFG_PLL_SOURCE, and BSP_CFG_PLL_DIV BSP will be used in the bsp_clock_init() to configure PLL clock.

Comment (gmarull) [reply dts/arm/renesas/ra/ra-cm4-common.dtsi:67]: FYI: We can't have HALisms in DT.

Comment (duynguyenxa) [reply dts/arm/renesas/ra/ra-cm4-common.dtsi:67]: @gmarull , the macro is defined in zephyr main repo, zephyr/include/zephyr/dt-bindings/clock/ra_clock.h

Comment (gmarull) [reply dts/arm/renesas/ra/ra-cm4-common.dtsi:67]: it still looks wrong, why isn't `div = <2>`? Are such values encoding what HAL expects?

Comment (duynguyenxa) [reply dts/arm/renesas/ra/ra-cm4-common.dtsi:67]: @gmarull , sorry for late response, it's expect the value map with HWM here for the pll.
image

File: soc/renesas/ra/ra4m1/Kconfig [hunk 3]
```diff
@@ -5,10 +5,14 @@ config SOC_SERIES_RA4M1
...
 	select CPU_CORTEX_M4
 	select CPU_HAS_ARM_MPU
+	select CPU_HAS_FPU
 	select CPU_CORTEX_M_HAS_SYSTICK
 	select DYNAMIC_INTERRUPTS
 	select TIMER_READS_ITS_FREQUENCY_AT_RUNTIME
 	select XIP
+	select HAS_RENESAS_RA_FSP
+	help
+	  Enable support for Renesas RA4M1 MCU series
```
Comment (nordicjm) [inline soc/renesas/ra/ra4m1/Kconfig:15]:
```suggestion
```

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: dts/arm/renesas/ra/ra-cm4-common.dtsi (hunk 1 above)
Comment (soburi) [reply dts/arm/renesas/ra/ra-cm4-common.dtsi:67]: @duynguyenxa 

The root of this problem is that the DeviceTree is a hardware description language,
and how values ​​are handled should not be mixed in here.
In other words, the DeviceTree should be able to describe semantic values; for example, if the division factor is 2, then it should be describable as "2", and it should be up to the implementation side, i.e., the .h or .c file, to convert this to bit notation.

Comment (soburi) [reply dts/arm/renesas/ra/ra-cm4-common.dtsi:67]: I splitted out this issue to #78365.

**Review by thaoluonguw — DISMISSED** [review]

**Comment by soburi** [comment]
@iandmorris I am looking to make part of the RA4M1 implementation FSP-based. Could you please review and check on your boards.

File: dts/arm/renesas/ra/ra-cm4-common.dtsi (hunk 2 above)
Comment (soburi) [reply dts/arm/renesas/ra/ra-cm4-common.dtsi]: https://github.com/zephyrproject-rtos/zephyr/pull/78097 is already merged.

File: soc/renesas/ra/ra4m1/Kconfig (hunk 3 above)
Comment (soburi) [reply soc/renesas/ra/ra4m1/Kconfig]: Fixed it.

**Review by nordicjm — DISMISSED** [review]
pinctrl and soc Kconfig/cmake changes OK

**Review by thaoluonguw — DISMISSED** [review]

**Comment by iandmorris** [comment]
> @iandmorris I am looking to make part of the RA4M1 implementation FSP-based. Could you please review and check on your boards.

@soburi - just confirmed that blinky still builds and runs on the Mikroe Clicker RA4M1

**Comment by soburi** [comment]
> > @iandmorris I am looking to make part of the RA4M1 implementation FSP-based. Could you please review and check on your boards.
> 
> @soburi - just confirmed that blinky still builds and runs on the Mikroe Clicker RA4M1

Thank you for your confirmation. I would like to proceed with this.

**Comment by soburi** [comment]
@KhiemNguyenT 
Could you take a look, please?

File: dts/arm/renesas/ra/ra-cm4-common.dtsi (hunk 1 above)
Comment (soburi) [reply dts/arm/renesas/ra/ra-cm4-common.dtsi]: https://github.com/zephyrproject-rtos/zephyr/pull/78365 is already merged.

**Review by KhiemNguyenT — DISMISSED** [review]

**Comment by soburi** [comment]
@gmarull 
Could you please re-check your change request?

File: soc/renesas/ra/ra4m1/soc.c [hunk 4]
```diff
@@ -131,6 +136,28 @@ const struct opt_set_mem ops __attribute__((section(".opt_set_mem"))) = {
...
+static int renesas_ra4m1_init(void)
+{
+	SystemCoreClock = BSP_MOCO_HZ;
+	g_protect_pfswe_counter = 0;
+	bsp_clock_init();
+
+	return 0;
+}
+
+SYS_INIT(renesas_ra4m1_init, PRE_KERNEL_1, 0);
```
Comment (gmarull) [inline soc/renesas/ra/ra4m1/soc.c:163]: use hooks

Comment (soburi) [reply soc/renesas/ra/ra4m1/soc.c]: I fixed it.
Thank you for your review.

**Review by gmarull — DISMISSED** [review]

**Review by thaoluonguw — DISMISSED** [review]
@soburi : It looks good to me for now.

File: boards/arduino/uno_r4/arduino_uno_r4_wifi-pinctrl.dtsi [hunk 5]
```diff
@@ -4,12 +4,18 @@
...
 
-#include <zephyr/dt-bindings/pinctrl/renesas/pinctrl-r7fa4m1xxxxxx.h>
+#include <zephyr/dt-bindings/pinctrl/renesas/pinctrl-ra.h>
 
 &pinctrl {
 	sci2_default: sci2_default {
 		group1 {
-			pinmux = <P301_RXD2>, <P302_TXD2>;
+			/* tx */
+			psels = <RA_PSEL(RA_PSEL_SCI_2, 3, 1)>;
```
[the remark this answers is not in the document]
Comment (soburi) [reply boards/arduino/uno_r4/arduino_uno_r4_wifi-pinctrl.dtsi:13]: Thank you. That's not good. I'll check it out.

[the remark this answers is not in the document]
Comment (soburi) [reply boards/arduino/uno_r4/arduino_uno_r4_wifi-pinctrl.dtsi]: @thaoluonguw 

I rechecked it.
As you pointed out, this doesn't seem right, so I fixed it.

Here, RX  and TX were only swapped incorrectly.
So, both are the same configuration set to SCI2 function and do not behave incorrectly.
So, I failed to catch in the test.

**Review by nordicjm — DISMISSED** [review]

**Review by thaoluonguw — DISMISSED** [review]

**Review by duynguyenxa — DISMISSED** [review]

**Comment by soburi** [comment]
@KhiemNguyenT
Please approve if it is good for you.

File: soc/renesas/ra/ra4m1/CMakeLists.txt [hunk 6]
```diff
@@ -3,7 +3,7 @@
 
 zephyr_include_directories(.)
 
-zephyr_library_sources_ifdef(CONFIG_SOC_OPTION_SETTING_MEMORY
+zephyr_library_sources(
```
Comment (tejlmand) [inline soc/renesas/ra/ra4m1/CMakeLists.txt:6]: minor nit, not blocking.

No need for linewrap anymore.

**Review by tejlmand — DISMISSED** [review]
lgtm, a minot nit observed but not blocking.

File: soc/renesas/ra/ra4m1/CMakeLists.txt (hunk 6 above)
Comment (soburi) [reply soc/renesas/ra/ra4m1/CMakeLists.txt]: Thank you for your feedback.
I fixed it.

**Review by thaoluonguw — APPROVED** [review]

**Review by nordicjm — APPROVED** [review]

**Comment by soburi** [comment]
@KhiemNguyenT 
Please take a look.

**Review by duynguyenxa — APPROVED** [review]

**Review by KhiemNguyenT — APPROVED** [review]
Thank you for this action.
LGTM.

**Pull request merged**